clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+        }
+        self.add_diagnostic(PlacedError {
+            span: Span {
+                start_line,
+                start_column,
+                end_line: self.last_line as u32,
+                end_column: self.last_column as u32,
+            },
+            e: err,
+        });
+    }
+
+    fn read_identifier(&mut self, first: Option<char>) -> String {
+        let mut ident = String::new();
+        if let Some(first) = first {
+            ident.push(first);
+        }
+
+        loop {
+            match self.next {
+                'a'..='z'
+                | 'A'..='Z'
+                | '0'..='9'
+                | '_'
+                | '-'
+                | '!'
+                | '?'
+                | '+'
+                | '<'
+                | '>'
+                | '='
+                | '/'
+                | '*' => ident.push(self.next),
+                _ => {
+                    if is_separator(self.next) {
+                        return ident;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharIdent(self.next));
+                        return ident;
+                    }
+                }
+            }
+            self.read_char();
+        }
+    }
+
+    fn read_trait_identifier(&mut self) -> String {
+        let start_line = self.last_line as u32;
+        let start_column = self.last_column as u32;
+        let mut ident = String::new();
+
+        loop {
+            match self.next {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '-' => ident.push(self.next),
+                '>' => return ident,
+                _ => {
+                    if is_separator(self.next) {
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line: self.line as u32,
+                                start_column: self.column as u32,
+                                end_line: self.line as u32,
+                                end_column: self.column as u32,
+                            },
+                            e: LexerError::ExpectedClosing('>'),
+                        });
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line,
+                                start_column,
+                                end_line: start_line,
+                                end_column: start_column,
+                            },
+                            e: LexerError::NoteToMatchThis('<'),
+                        });
+                        return ident;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharTraitIdent(self.next));
+                        return ident;
+                    }
+                }
+            }
+            self.read_char();
+        }
+    }
+
+    fn read_principal(&mut self) -> String {
```

Comment:
Ah, I think this code just reads `X` in `'X`, right?  It doesn't really care whether or not `X` represents a valid principal?

Reply:
Yes, this is just reading `X` in `'X`.

Right, the parser checks that using `parse_standard_principal`, but I had not realized that principals use the Crockford 32 alphabet. Maybe the lexer should catch that, since it would be able to give a more specific error in the case that an illegal character is used. I will modify the checks there.

Reply:
Fixed in 2d825c5ba